Enhance socket documentation

jukkar opened the pull request:
Add more documentation groups for socket.h

I was not able to build the html documentation locally so I need to submit the PR multiple times in order to see what it looks like.

jukkar committed:
net: socket: Group various defines together for documentation

Groupped various defines in socket.h together based on their
usage. This groupping is only for documentation.

Fixes #66081

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: TLS doc group was missing defines

Couple of TLS defines were not in correct documentation group.
Add also missing doxygen comments for TLS defines.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

kartben commented on include/zephyr/net/socket.h:
```diff
@@ -41,6 +41,10 @@ struct zsock_pollfd {
 };
 
+/**
+ * @defgroup poll_options Options for poll()
```
I think you might be looking at using `@name` groups, rather than `@defgroup`. The former is more geared towards semantically grouping stuff together, whereas the latter ends up calling the group "modules", which is probably more than what you need? Added a task to #6810 so that I document this as a guideline at some point :)

jukkar replied:
Thanks for the hints! I will change to use `@name` groups then.

jukkar commented:
The documentation is here https://builds.zephyrproject.io/zephyr/pr/66108/docs/connectivity/networking/api/sockets.html
The socket docs needs still more work.

jukkar committed:
net: socket: Update socket option documentation

Remove redundant information from socket option documentation.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Add missing doxygen documentation

Add doxygen comments to those APIs where they were missing
and were relevant.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Remove "ignored" comment from supported symbols

Some socket symbols were marked as "ignored" even when they
are in fact supported.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on include/zephyr/net/socket.h:
```diff
@@ -203,48 +222,57 @@ struct zsock_pollfd {
+	int ai_socktype;          /**< Socket type, for example SOCK_STREAM or SOCK_DGRAM */
+	int ai_protocol;          /**< Protocol for addresses, 0 means any protocol */
+	socklen_t ai_addrlen;     /**< Length of the socket address */
+	struct sockaddr *ai_addr; /** Pointer to the address */
```
```suggestion
	struct sockaddr *ai_addr; /**< Pointer to the address */
```

rlubos commented on include/zephyr/net/socket.h:
```diff
@@ -1273,7 +1273,7 @@ struct in6_pktinfo {
  * @name Backlog size for listen()
  * @{
  */
-/** listen: The maximum backlog queue length (ignored, for compatibility) */
```
Shall this be removed? We don't implement this option

jukkar replied:
It is actually not an option, but just a define telling the maximum backlog length. I think we can leave it here, but we could/should probably adjust the value a bit. It seems to be quite long.

jukkar commented on include/zephyr/net/socket.h:
```diff
@@ -203,48 +222,57 @@ struct zsock_pollfd {
 #define TLS_PEER_VERIFY_OPTIONAL 1 /**< Peer verification optional. */
 #define TLS_PEER_VERIFY_REQUIRED 2 /**< Peer verification required. */
 
-/* Valid values for TLS_DTLS_ROLE option */
+/** Valid values for TLS_DTLS_ROLE option */
 #define TLS_DTLS_ROLE_CLIENT 0 /**< Client role in a DTLS session. */
 #define TLS_DTLS_ROLE_SERVER 1 /**< Server role in a DTLS session. */
 
-/* Valid values for TLS_CERT_NOCOPY option */
+/** Valid values for TLS_CERT_NOCOPY option */
 #define TLS_CERT_NOCOPY_NONE 0     /**< Cert duplicated in heap */
 #define TLS_CERT_NOCOPY_OPTIONAL 1 /**< Cert not copied in heap if DER */
 
-/* Valid values for TLS_SESSION_CACHE option */
+/** Valid values for TLS_SESSION_CACHE option */
 #define TLS_SESSION_CACHE_DISABLED 0 /**< Disable TLS session caching. */
 #define TLS_SESSION_CACHE_ENABLED 1 /**< Enable TLS session caching. */
 
-/* Valid values for TLS_DTLS_CID option */
+/** Valid values for TLS_DTLS_CID option */
 #define TLS_DTLS_CID_DISABLED		0
 #define TLS_DTLS_CID_SUPPORTED		1
 #define TLS_DTLS_CID_ENABLED		2
 
-/* Valid values for TLS_DTLS_CID_STATUS option */
+/** Valid values for TLS_DTLS_CID_STATUS option */
 #define TLS_DTLS_CID_STATUS_DISABLED		0
 #define TLS_DTLS_CID_STATUS_DOWNLINK		1
 #define TLS_DTLS_CID_STATUS_UPLINK		2
 #define TLS_DTLS_CID_STATUS_BIDIRECTIONAL	3
```
Ok, those settings were inside defgroup so I missed the additional @name.

kartben commented on include/zephyr/net/socket.h:
```diff
@@ -203,48 +222,57 @@ struct zsock_pollfd {
 #define TLS_CERT_NOCOPY_OPTIONAL 1 /**< Cert not copied in heap if DER */
 
-/* Valid values for TLS_SESSION_CACHE option */
+/** Valid values for TLS_SESSION_CACHE option */
```
might make sense to make occurences like these actual `@ref`'s

```suggestion
/** Valid values for @ref TLS_SESSION_CACHE option */
```

kartben replied:
I apologize as it looks like I posted several comments for the same line that had conflicting "suggested changes" :(

You can't have this be a "regular" doxygen comment -- the way it is currently it applies to `TLS_PEER_VERIFY_NONE` https://builds.zephyrproject.io/zephyr/pr/66108/docs/doxygen/html/group__secure__sockets__options.html#ga09cb746907891d86a8d69ca49717c068

Instances of such comments should either be part of a `@name`, or turned into simple `/*  ...  */` so that they are here for reference in the code itself but don't show up in the doxygen docs.

FWIW, since nesting @name is not allowed, maybe `@name Socket options for TLS` should be a `@defgroup` after all (and then added to sockets.rst as a `doxygengroup:: ...`).

kartben replied:
@jukkar see previous comment that you may have missed (and apologies again if my previous "suggested changes" might have been confusing).

jukkar replied:
I indeed missed what changes I should do. And after reading your comment again and looking the generated output my wooden eyes still don't see what changes you would like to see here. Could you elaborate more?
Note that there is no nested `@name` in socket.h in this PR.

kartben reviewed (changes requested):
BTW you might want to use https://builds.zephyrproject.io/zephyr/pr/66108/api-coverage/zephyr/net/socket.h.gcov.html as a way to quickly visualize some of the missing comments.
Thanks for this work!

jukkar commented:
Missed latest comments as I pushed new version at the same time. Will send at least one more version.

jukkar commented:
> you might want to use https://builds.zephyrproject.io/zephyr/pr/66108/api-coverage/zephyr/net/socket.h.gcov.html as a way to quickly visualize some of the missing comments.

Nice, I did not know about that link. Thanks for the info :+1:

jukkar commented:
@kartben ping, if you have time, please re-visit this one.

jukkar commented:
@kartben Could you elaborate what changes are still needed, I could not figure out what to do from the comment.

jukkar commented:
@kartben did you meant this with the TLS comments you had earlier

https://builds.zephyrproject.io/zephyr/pr/66108/docs/doxygen/html/group__secure__sockets__options.html#gace333e12f9d74f1ff7c5ac71f7facd16

so only `TLS_PEER_VERIFY_NONE` has a link to `TLS_PEER_VERIFY` option, but the other `required` and `optional` options do not.
I think it would be better to have the options documented so I would keep the `/**< ` comments and optionally remove the `@ref` although I don't think the missing two ref links look too bad either.

kartben commented:
> @kartben did you meant this with the TLS comments you had earlier
> 
> https://builds.zephyrproject.io/zephyr/pr/66108/docs/doxygen/html/group__secure__sockets__options.html#gace333e12f9d74f1ff7c5ac71f7facd16
> 
> so only `TLS_PEER_VERIFY_NONE` has a link to `TLS_PEER_VERIFY` option, but the other `required` and `optional` options do not. I think it would be better to have the options documented so I would keep the `/**< ` comments and optionally remove the `@ref` although I don't think the missing two ref links look too bad either.

Sorry that I wasn't clear. What I mean is that by having the below, you're effectively having the first line be a doxygen comment applying to ``TLS_PEER_VERIFY_NONE`` (and only it, just look where "Valid values for TLS_PEER_VERIFY option." ends up in the generated doc)

```
/** Valid values for @ref TLS_PEER_VERIFY option */
#define TLS_PEER_VERIFY_NONE 0     /**< Peer verification disabled. */
#define TLS_PEER_VERIFY_OPTIONAL 1 /**< Peer verification optional. */
#define TLS_PEER_VERIFY_REQUIRED 2 /**< Peer verification required. */
```

The "ideal" solution would be to create a `@name` section for these 3 options (and of course the couple other similar groups you have in the file) but this won't work since it's not possible to nest a `@name` group in the existing `@name 
Socket options for TLS` one we're already in.

So either  just make this a simple C comment (`/* Valid values for @ref TLS_PEER_VERIFY option */`) to have the grouping at least described for folks browsing the code ; or you could create a doxygen group for all TLS options, and then have named sections in it?
HTH!

jukkar commented:
> What I mean is that by having the below, you're effectively having the first line be a doxygen comment applying to `TLS_PEER_VERIFY_NONE` (and only it, just look where "Valid values for TLS_PEER_VERIFY option." ends up in the generated doc)

Yep, I realized what you meant after looking the latest generated documentation.

I will send a new version that removes the doxygen comment from where was a ref. We can add additional groups later if really needed.
